sycl/tools/sycl-ls/sycl-ls.cpp

```diff
@@ -0,0 +1,157 @@
+// how clinfo prints this for OpenCL devices.
+//
+// There are two types of output:
+//   concise (default) and
+//   verbose (enabled with --verbose).
+//
+// In verbose mode it also prints, which devices would be chosen by various SYCL
+// device selectors.
+//
+#include <CL/sycl.hpp>
+
+#include <cstdlib>
+#include <iostream>
+#include <stdlib.h>
+
+using namespace cl::sycl;
+
+// Controls verbose output vs. concise.
+bool verbose;
+
+// Trivial custom selector that selects a device of the given type.
+class custom_selector : public device_selector {
+  info::device_type MType;
+
+public:
+  custom_selector(info::device_type Type) : MType(Type) {}
+  int operator()(const device &Dev) const override {
+    return Dev.get_info<info::device::device_type>() == MType ? 1 : -1;
+  }
+};
+
+static void printDeviceInfo(const device &Device, const std::string &Prepend) {
+  auto DeviceType = Device.get_info<info::device::device_type>();
+  std::string DeviceTypeName;
+  switch (DeviceType) {
+  case info::device_type::cpu:
+    DeviceTypeName = "CPU ";
+    break;
+  case info::device_type::gpu:
+    DeviceTypeName = "GPU ";
+    break;
+  case info::device_type::host:
+    DeviceTypeName = "HOST";
+    break;
+  case info::device_type::accelerator:
+    DeviceTypeName = "ACC ";
+    break;
+  default:
+    DeviceTypeName = "UNKNOWN";
+    break;
+  }
+
+  auto DeviceVersion = Device.get_info<info::device::version>();
+  auto DeviceName = Device.get_info<info::device::name>();
+  auto DeviceVendor = Device.get_info<info::device::vendor>();
+  auto DeviceDriverVersion = Device.get_info<info::device::driver_version>();
+
+  if (verbose) {
+    std::cout << Prepend << "Type       : " << DeviceTypeName << std::endl;
+    std::cout << Prepend << "Version    : " << DeviceVersion << std::endl;
+    std::cout << Prepend << "Name       : " << DeviceName << std::endl;
+    std::cout << Prepend << "Vendor     : " << DeviceVendor << std::endl;
+    std::cout << Prepend << "Driver     : " << DeviceDriverVersion << std::endl;
+  } else {
+    std::cout << Prepend << DeviceTypeName << ": " << DeviceVersion << "[ "
+              << DeviceDriverVersion << " ]" << std::endl;
+  }
+}
+
+static void printSelectorChoice(const device_selector &Selector,
+                                const std::string &Prepend) {
+  try {
+    const auto &Dev = device(Selector);
+    printDeviceInfo(Dev, Prepend);
+
+  } catch (const cl::sycl::runtime_error &Exception) {
+    // Truncate long string so it can fit in one-line
+    std::string What = Exception.what();
+    if (What.length() > 50)
+      What = What.substr(0, 50) + "...";
+    std::cout << Prepend << What << std::endl;
+  }
+}
+
+int main(int argc, char **argv) {
+
+  // See if verbose output is requested
+  if (argc == 1)
+    verbose = false;
+  else if (argc == 2 && std::string(argv[1]) == "--verbose")
+    verbose = true;
+  else {
+    std::cout << "Usage: sycl-ls [--verbose]" << std::endl;
+    return -1;
```

Comment:
Very minor comment, but this should be `EXIT_FAILURE`

Reply:
OK